Support Kolibri 0.14 and use the app plugin

dylanmccall opened the pull request:
To support this, import Kolibri directly instead of using subprocess. As a result, the content extensions work from the Kolibri flatpak needs to be moved here as well.

This probably looks somewhat insane, so some sanity checking is very appreciated. The idea behind the `kolibri_service` module and the split up `globals` module is that whenever we import stuff from Kolibri, there tend to be side-effects.

So by isolating stuff into processes, which are separated more strictly than Python threads, I found it easier to manage that. It also gave me an opportunity to draw some lines where we can untangle various interdependencies in the future.

dnohales commented on src/kolibri_gnome/desktop_launcher/application.py:
```diff
@@ -42,7 +50,8 @@ def on_close_window(self):
     def on_open_in_browser(self):
-        subprocess.call(["xdg-open", self.get_current_or_target_url()])
+        current_url = self.get_current_or_target_url()
+        subprocess.call(["xdg-open", current_url])
```
This is not needed, I actually prefer the previous style :P

dnohales commented on src/kolibri_gnome/desktop_launcher/application.py:
```diff
@@ -261,14 +265,15 @@ def init_ui(self):
-        if "URL" in saved_state and saved_state["URL"].startswith(KOLIBRI_URL):
-            pew.ui.run_on_main_thread(main_window.load_url, saved_state["URL"])
+        saved_url = saved_state.get("URL")
+        if saved_url and saved_url.startswith(KOLIBRI_URL):
```
You can use `is_kolibri_url` here.

dnohales commented on src/kolibri_gnome/desktop_launcher/application.py:
```diff
@@ -277,73 +282,49 @@ def __init_service(self):
-        self.__kolibri_run_thread = pew.ui.PEWThread(target=self.run_server)
-        self.__kolibri_run_thread.daemon = False
-        self.__kolibri_run_thread.start()
 
```
Remove this blank line

dylanmccall commented:
Thanks for the review @dnohales ! I took care of those things :)

jamalex commented on src/kolibri_gnome/kolibri_service/kolibri_service_main.py:
```diff
@@ -0,0 +1,111 @@
+                else:
+                    return None
+
+        logger.info("Kolibri is not starting. Giving up.")
```
![image](https://user-images.githubusercontent.com/618416/90080203-df82a700-dcbe-11ea-8047-60e143c43488.png)

jamalex replied:
Joking aside: I don't see "max tries" logic or anything here. The only place I see the event being cleared is when `stop` is called, or maybe I'm misreading? So is this really "giving up"?

dylanmccall replied:
Oh yeah, that is definitely dead code. (I think it runs on a clean exit, actually). Good catch!

jamalex commented on src/kolibri_gnome/kolibri_service/kolibri_service_monitor.py:
```diff
@@ -0,0 +1,46 @@
+        while not is_kolibri_responding():
+            # There is a corner case here where Kolibri may be running (lock
+            # file is created), but responding at a different URL than we
+            # expect. This is unlikely, so we are ignoring it here.
```
Curious: Is that related to the path prefix setting noted above? Or are you referring to something else?

dylanmccall replied:
Now you mention it, I think that locking functionality can be removed at this point, seeing as we're directly importing Kolibri which handles its own locking just fine. I have removed the extra lock file, and thus that corner case :)

jamalex commented on src/kolibri_gnome/kolibri_service/kolibri_service_setup.py:
```diff
@@ -0,0 +1,153 @@
+
+from .content_extensions import ContentExtensionsList
+
+KOLIBRI = "kolibri"
```
Checks out. :+1:

jamalex commented on src/kolibri_gnome/kolibri_service/content_extensions.py:
```diff
@@ -0,0 +1,329 @@
+import itertools
```
Ah, I'm not sure why I thought it wasn't imported into other places in the project here, but I see that it is. Do you have any ideas for how it might be a bit more modularized? Maybe something along the lines of having a hook or registry system of some sort where the content extension stuff injects itself so that it's then called at the right time in the app lifecycle...?

(We could open this as a tech debt issue if we want, to be dealt with as part of the "cross-OS app reconciliation process")

dylanmccall replied:
My thought here is the Flatpak-specific chunk of it is pretty minimal. Right now it uses commit hashes from /.flatpak-info to detect version changes and to find the list of installed extensions. But if we wanted to support more general-purpose content extensions we'd probably end up with our own ID format and we'd look through some well-known directory. So no more flatpak-info parsing, and then the only part that's really special is going to be file paths. (To be honest, that would be better anyway but it's just a smidge more work). I'll add a docstring explaining where it sits right now.

dylanmccall commented on src/kolibri_gnome/desktop_launcher/application.py:
```diff
@@ -208,10 +221,6 @@ def __init__(self, *args, **kwargs):
-        # Set a user agent so Kolibri behaves differently
-        # <https://github.com/learningequality/kolibri/blob/app-support/kolibri/core/assets/src/utils/browser.js#L25>
-        user_agent = self.gtk_webview.get_settings().get_user_agent()
-        self.set_user_agent(" ".join([user_agent, "Android", "wv", "KolibriApp/1.0"]))
```
Yeah, from what I can tell enabling the app plugin does what I was wanting there - simplified setup wizard (or is that everywhere now?) and user selector at the login screen. I'm not super familiar with that, though, so correct me if wrong :)

dylanmccall commented on src/kolibri_gnome/desktop_launcher/application.py:
```diff
@@ -261,89 +263,57 @@ def init_ui(self):
+            raise LoadRedirectError()
+        elif self.__kolibri_service_manager.is_kolibri_url(target_url):
+            root_url = self.__kolibri_service_manager.get_initialize_url(target_url)
+            raise LoadRedirectURL(root_url)
```
Looking at it again, that last case should totally be returning a value! I changed this to throw an exception for the error and loading redirects (since those are somewhat exceptional cases), and return a value for the regular case. That feels more Pythonic although I can't recall the specific idiom I'm going with there :)

dylanmccall commented on src/kolibri_gnome/kolibri_globals.py:
```diff
@@ -0,0 +1,56 @@
+else:
+    from kolibri.utils.conf import OPTIONS
+
+    KOLIBRI_HTTP_PORT = OPTIONS["Deployment"]["HTTP_PORT"]
```
Oh yeah, that is a good one. I cleaned this up a bunch in a new commit.

dylanmccall commented on src/kolibri_gnome/kolibri_globals.py:
```diff
@@ -0,0 +1,56 @@
+
+    KOLIBRI_HTTP_PORT = OPTIONS["Deployment"]["HTTP_PORT"]
+
+KOLIBRI_URL = "http://localhost:{}".format(KOLIBRI_HTTP_PORT)
```
Good point! That should be working now. I took the opportunity to tidy up KOLIBRI_URL (now KOLIBRI_BASE_URL) in kolibri_globals, removing KOLIBRI_URL_SPLIT and just splitting it where appropriate. The `_replace` logic was indeed a bit ugly, but only in two places :man_shrugging:

dylanmccall commented on src/kolibri_gnome/kolibri_globals.py:
```diff
@@ -0,0 +1,56 @@
+    except KolibriAPIError:
+        return False
+    else:
+        return info.get("application") == "kolibri"
```
Done! Added an `if isinstance(info, collections.Mapping)`.

dylanmccall commented:
I worked through those details, added some doc strings, and removed some unnecessary lines of code. All good if we go ahead and merge this today? :) Then we can get the flatpak changes organized.
